ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBSynthesizer/Run.hs

```diff
@@ -0,0 +1,163 @@
+                     (nodeImmutableDbChunkInfo)
+import           Ouroboros.Consensus.Node.ProtocolInfo (ProtocolInfo (..))
+import           Ouroboros.Consensus.Shelley.Node (ShelleyGenesis (..),
+                     validateGenesis)
+import qualified Ouroboros.Consensus.Storage.ChainDB as ChainDB (defaultArgs)
+import qualified Ouroboros.Consensus.Storage.ChainDB.Impl as ChainDB (cdbTracer,
+                     withDB)
+import           Ouroboros.Consensus.Storage.LedgerDB.DiskPolicy
+                     (SnapshotInterval (..), defaultDiskPolicy)
+import           Ouroboros.Consensus.Util.ResourceRegistry
+
+import           Cardano.Api.Any (displayError)
+import           Cardano.Api.Protocol.Types (protocolInfo)
+import           Cardano.Node.Protocol
+import           Cardano.Node.Types
+
+import           Control.Monad (unless)
+import           Data.Bool (bool)
+import           System.Directory
+import           System.FilePath (takeDirectory, (</>))
+
+import           Data.Aeson as Aeson (FromJSON, Result (..), Value,
+                     eitherDecodeFileStrict', eitherDecodeStrict', fromJSON)
+import           Data.ByteString as BS (ByteString, readFile)
+
+import           Control.Monad.Trans.Except (ExceptT)
+import           Control.Monad.Trans.Except.Extra (firstExceptT,
+                     handleIOExceptT, hoistEither, runExceptT)
+
+import           Control.Tracer (nullTracer)
+{-
+import           Control.Monad.IO.Class
+import           GHC.Clock
+import           System.IO
+-}
+
+
+initialize
+    :: NodeFilePaths
+    -> NodeCredentials
+    -> DBSynthesizerOptions
+    -> IO (Either String (DBSynthesizerConfig, SomeConsensusProtocol))
+initialize NodeFilePaths{..} NodeCredentials{..} synthOptions = do
+    relativeToConfig :: (FilePath -> FilePath) <-
+        (</>) . takeDirectory <$> makeAbsolute nfpConfig
+    runExceptT $ do
+        conf    <- initConf relativeToConfig
+        proto   <- initProtocol relativeToConfig conf
+        pure    (conf, proto)
+  where
+    initConf :: (FilePath -> FilePath) -> ExceptT String IO DBSynthesizerConfig
+    initConf relativeToConfig = do
+        inp                 <- handleIOExceptT show (BS.readFile nfpConfig)
+        confConfigStub      <- adjustFilePaths relativeToConfig <$> readJson inp
+        confShelleyGenesis  <- readFileJson $ ncsShelleyGenesisFile confConfigStub
+        _                   <- hoistEither $ validateGenesis confShelleyGenesis
+        let
+            confProtocolCredentials = ProtocolFilepaths {
+              byronCertFile         = Nothing
+            , byronKeyFile          = Nothing
+            , shelleyKESFile        = credKESFile
+            , shelleyVRFFile        = credVRFFile
+            , shelleyCertFile       = credCertFile
+            , shelleyBulkCredsFile  = credBulkFile
+            }
+        pure DBSynthesizerConfig {
+              confDbDir     = nfpChainDB
+            , confOptions   = synthOptions
+            , ..
```

Comment:
I'm trying to vanish wildcards from the code base. I wonder if we want to avoid them here as well.